add option to enable or disable from default metrics

aanm opened the pull request:
This PR adds metric levels which allows for users to specify which metrics they want to choose. Right now the bpf syscalls metrics will be the only ones not considered the default.

```release-note
Add option to enable or disable from default metrics (`+metric_foo` to enable `metric_foo`, `-metric_bar` to disable `metric_bar`)
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7896)
<!-- Reviewable:end -->

aanm committed:
pkg/option: add metrics option to enable or disable from default metrics

This gives the ability to enable or disable metrics in cilium from the
given default list of metrics in Cilium.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/bpf: only account for bpf syscalls if syscall metric is enabled

This will make bpf syscalls to not be accounted by default as they
consume lots of memory allocations.

Signed-off-by: André Martins <andre@cilium.io>

tgraf commented on daemon/daemon_main.go:
```diff
@@ -602,6 +602,9 @@ func init() {
 	flags.MarkHidden(option.MaxCtrlIntervalName)
 	option.BindEnv(option.MaxCtrlIntervalName)
 
+	flags.Int(option.MetricsLevel, option.MetricLevelCore, "Level of metrics that Cilium exposes.")
```
What about we  do a StringSlice and allow to define a mask of what should be recorded?

`--metrics=kvstore-histogram --metrics=bpf-syscall [...]`

ianvernon replied:
Agreed, this is more user-friendly, as the choices for integers are arbitrary.

joestringer commented on pkg/option/config.go:
```diff
@@ -442,6 +442,22 @@ const (
+const (
+	// MetricLevelNone should not expose any metric!
+	MetricLevelNone int = iota
+	// MetricLevelCore should expose core TBD
```
core TBD?

aanm replied:
core, To Be Decided :)

aanm committed:
pkg/metrics: set subsystems and labels as constants

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
daemon: use constant SubsystemAgent from pkg/metrics

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/kvstore: disable metric collection if KVStore metrics are not enabled

Signed-off-by: André Martins <andre@cilium.io>

ianvernon commented on daemon/daemon_main.go:
```diff
@@ -602,6 +602,9 @@ func init() {
 	flags.MarkHidden(option.MaxCtrlIntervalName)
 	option.BindEnv(option.MaxCtrlIntervalName)
 
+	flags.StringSlice(option.SubsysMetrics, metrics.DefaultSubsystems(), "Subsystems metrics that Cilium exposes.")
```
remove the period

ianvernon commented on pkg/bpf/bpf_linux.go:
```diff
@@ -128,14 +139,19 @@ func LookupElement(fd int, key, value unsafe.Pointer) error {
 
-	duration := spanstat.Start()
+	var duration *spanstat.SpanStat
+	if option.Config.IsSubsysMetricEnabled(metrics.SubsystemBPFMask) {
```
This isn't blocking, but I feel like this check would be way simpler overall if the value being checked was a variable set at the package level? We can use `option.Config` to gather all configuration from daemon bootstrap, then at bootstrap of this package, take in the value from `option.Config` and set it as a package-scoped variable.

aanm replied:
@ianvernon That's going definitely going to be a follow up as the performance improves as well.

```
5506 ns/op vs 5365 ns/op
```

ianvernon reviewed (approved):
I have no blocking comments - it makes sense to get this merged sooner rather than later so we can get it into `v1.5`. I do think that it would be better for code readability to set the metrics configuration as a package-level variable from `pkg/option` instead of having to call back into `option.Config` every time we need to do a lookup.

tgraf commented on pkg/metrics/metrics.go:
```diff
@@ -55,6 +56,98 @@ const (
+
+// DefaultSubsystems returns the default subsystems that will be used to expose
+// metrics.
+func DefaultSubsystems() []string {
```
I don't think the granularity of entire subsystems will be sufficient. Most subsystems will be scalable and less scalable metrics and users are likely to opt out of less scalable metrics.

I see two paths that make sense:

1. High-level flag that disables less scalable metrics such as histograms  
2. A per metric systems where we have a default `[]string{metric1,metric2,metric3}` and then allow to modify that with `--metrics=+metric2 --metrics=-metric3` to allow enabling or disabling individual metrics without requiring to be aware of the default all the time.

aanm replied:
@tgraf I hope that I'm not understanding 2) correctly, we will give the users to select over 50 different metrics, this sounds extremely overly complicated. If users don't want those metrics from being considered in prometheus they can configure prometheus to do that.

tgraf replied:
How does your current approach solve the problem of providing metrics while being able to disable the ones that are not scalable? I don't think we want users to get too involved. We want to have a good set of defaults and allow developers and advanced to enable individual metrics. We can also to `+grp-XXX` to enable individual groups.

I think subsystem is an arbitrary grouping that will be incredibly limiting if users want to find a good trade-off between visibility and load.

aanm replied:
>How does your current approach solve the problem of providing metrics while being able to disable the ones that are not scalable?

Scalable where? In prometheus or Cilium? In Cilium the only non scalable metrics are the bpf related ones. In prometheus, users can specify which metrics they want to disable in its configuration.

tgraf commented on pkg/option/config.go:
```diff
@@ -965,6 +973,11 @@ func (c *DaemonConfig) IsFlannelMasterDeviceSet() bool {
 }
 
+// IsSubsysMetricEnabled returns true if the given subsystem metric is enabled.
+func (c *DaemonConfig) IsSubsysMetricEnabled(subsystem uint64) bool {
```
What's the benefit of a uint64 mask over a struct with booleans? It seems simpler to be able to:
```
if option.Config.EnabledMetrics.BpfSyscalls {
```

aanm committed:
pkg/metrics: add CounterVec and GaugeVec interfaces

As prometheus golang_library does not offer those interfaces we need to
add our own specification of those interfaces. This will allow to have
no-op implementations of those interfaces where they will be used when a
metric is disabled.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/metrics: use interfaces for all metrics

Having interfaces will allow the ability to use different
implementations for the same metric. For example, in case a user wants
to disable a particular metric, that metric will have a no-op
implementation of the given interface.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/metrics: add no-op implementations for disabled metrics

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
common: add MapStringStructToSlice function

This allows to convert a map[string]struct{} into a slice of string
which contains all keys from the given map.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/metrics: set all metrics as a no-op unless they are enabled

When initializing Cilium, all metrics will be set as a no-op.
Registration as well as the real implementation of those metrics will
only be enforced made once CreateConfiguration function is called with
the given metric names.

Signed-off-by: André Martins <andre@cilium.io>

joestringer commented on pkg/metrics/metrics.go:
```diff
@@ -557,72 +365,594 @@ var (
-	MustRegister(FQDNGarbageCollectorCleanedTotal)
+func init() {
+	MustRegister(prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{Namespace: Namespace}))
+	// TODO: Figure out how to put this into a Namespace
```
What's the story behind this?

aanm replied:
It was already there. git blames @raybejjani on this one

joestringer commented on pkg/metrics/interfaces.go:
```diff
@@ -0,0 +1,127 @@
+	return NoOpObserverVec
+}
+
+// Gauge
```
```suggestion
// gauge is a prometheus metric that acts as a no-op for use when metrics for subsystems are disabled.
```

aanm replied:
@joestringer I've added that to make it easier the separation of all structures. I believe they are all self-explanatory. I've added the following warning to the one that is not.

```golang
// *WARNING*: Desc returns nil so do not register this metric into prometheus
// default register.
func (m *metric) Desc() *prometheus.Desc  { return nil }
```

joestringer reviewed (approved):
BPF bits LGTM, One todo I noticed below and a minor nit.

tgraf reviewed (approved):
Awesome work!
